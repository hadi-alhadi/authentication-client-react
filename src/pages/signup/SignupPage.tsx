import React, { useEffect } from "react";
import { useNavigate } from "react-router-dom";
import { Formik } from "formik";
import { Card } from "react-bootstrap";
import Form from "react-bootstrap/Form";
import Col from "react-bootstrap/Col";
import InputGroup from "react-bootstrap/InputGroup";
import Button from "react-bootstrap/Button";
import { NotificationManager } from "react-notifications";
import * as yup from "yup";
import { useSignupMutation } from "../../services/auth.service";
import { setShowLoader } from "../../redux/slices/general.slice";
import { useAppDispatch } from "../../redux/hooks";
import "./signupPage.css";
import { setAuthenticatedUser } from "../../redux/slices/auth.slice";

const schema = yup.object().shape({
  name: yup.string().required("Name is required"),
  email: yup
    .string()
    .required("Email is required")
    .email("Invalid email address"),
  password: yup
    .string()
    .required("Password is required")
    .min(8, "Password must be at least 8 characters")
    .matches(/[a-zA-Z]/, "Password must contain at least one letter")
    .matches(/[0-9]/, "Password must contain at least one number")
    .matches(
      /[!@#$%^&*(),.?":{}|<>]/,
      "Password must contain at least one special character",
    ),
});

const SignupPage = () => {
  const [signup, { data, error, isLoading }] = useSignupMutation();
  const dispatch = useAppDispatch();
  const navigate = useNavigate();

  useEffect(() => {
    if (data && !error) {
      console.log("SignUpPage:: data:", data);
      NotificationManager.success(
        `Thank you ${data.name}`,
        "Registration Success",
      );
      localStorage.setItem("user", JSON.stringify(data));
      dispatch(setAuthenticatedUser(data));
      navigate("/home");
    } else if (error) {
      NotificationManager.error(
        "Error authenticating user, please check your email and password",
        "Authentication Error",
      );
      console.log(`SignupPage: Authentication error`, error);
    }
  }, [data, error, dispatch, navigate]);

  useEffect(() => {
    dispatch(setShowLoader(isLoading));
  }, [isLoading, dispatch]);

  const handleSignup = (formValue: {
    name: string;
    email: string;
    password: string;
  }) => {
    const { name, email, password } = formValue;
    signup({ name, email, password });
  };

  return (
    <div className="signup-wrapper">
      <Formik
        validationSchema={schema}
        onSubmit={handleSignup}
        initialValues={{
          name: "",
          email: "",
          password: "",
        }}
      >
        {({
          handleSubmit,
          handleChange,
          handleBlur,
          values,
          touched,
          errors,
        }) => (
          <Card className="card" style={{ width: "18rem" }}>
            <Card.Title className="title">Sign Up Page</Card.Title>
            <Form className="form" noValidate onSubmit={handleSubmit}>
              <Form.Group as={Col} md="12" controlId="validationFormikName">
                <Form.Label>Name</Form.Label>
                <InputGroup hasValidation>
                  <InputGroup.Text id="inputGroupPrepend">@</InputGroup.Text>
                  <Form.Control
                    type="text"
                    placeholder="Your Name"
                    aria-describedby="inputGroupPrepend"
                    name="name"
                    value={values.name}
                    onChange={handleChange}
                    isInvalid={!!errors.name && touched.name}
                    onBlur={handleBlur}
                  />
                  {touched.name && (
                    <Form.Control.Feedback type="invalid">
                      {errors.name}
                    </Form.Control.Feedback>
                  )}
                </InputGroup>
              </Form.Group>
              <Form.Group as={Col} md="12" controlId="validationFormikEmail">
                <Form.Label>Email</Form.Label>
                <InputGroup hasValidation>
                  <InputGroup.Text id="inputGroupPrepend">@</InputGroup.Text>
                  <Form.Control
                    type="text"
                    placeholder="Your email address"
                    aria-describedby="inputGroupPrepend"
                    name="email"
                    value={values.email}
                    onChange={handleChange}
                    isInvalid={!!errors.email && touched.email}
                    onBlur={handleBlur}
                  />
                  {touched.email && (
                    <Form.Control.Feedback type="invalid">
                      {errors.email}
                    </Form.Control.Feedback>
                  )}
                </InputGroup>
              </Form.Group>

              <Form.Group as={Col} md="12" controlId="validationFormikPassword">
                <Form.Label>Password</Form.Label>
                <InputGroup hasValidation>
                  <InputGroup.Text id="inputGroupPrepend">
                    &#128273;
                  </InputGroup.Text>
                  <Form.Control
                    type="password"
                    placeholder="Your Password"
                    aria-describedby="inputGroupPrepend"
                    name="password"
                    value={values.password}
                    onChange={handleChange}
                    isInvalid={!!errors.password && touched.password}
                    onBlur={handleBlur}
                  />
                  {touched.password && (
                    <Form.Control.Feedback type="invalid">
                      {errors.password}
                    </Form.Control.Feedback>
                  )}
                </InputGroup>
              </Form.Group>
              <div>
                <span>Already have an account?</span>{" "}
                <a href="/signin">Sign In</a>
              </div>
              <Button type="submit">Sign Up</Button>
            </Form>
          </Card>
        )}
      </Formik>
    </div>
  );
};

export default SignupPage;
