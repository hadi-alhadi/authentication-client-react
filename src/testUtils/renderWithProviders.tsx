import React from "react";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import configureStore from "redux-mock-store";

const mockStore = configureStore();
const store = mockStore({});
export default function renderWithProviders(props: React.ReactNode) {
  return {
    ...render(
      <Provider store={store}>
        <MemoryRouter>{props}</MemoryRouter>
      </Provider>,
    ),
    history,
  };
}
