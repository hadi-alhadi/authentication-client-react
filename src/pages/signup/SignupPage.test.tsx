import React from "react";
import { fireEvent, waitFor, screen } from "@testing-library/react";
import SignupPage from "./SignupPage";
import renderWithProviders from "../../testUtils/renderWithProviders"; // Adjust the import path as necessary

const mockSignupFn = jest.fn(() =>
  Promise.resolve({ data: { name: "Test User", email: "test@example.com" } }),
);
jest.mock("../../services/auth.service", () => ({
  authAPI: {
    useSignupMutation: () => [
      mockSignupFn,
      { isLoading: false, isSuccess: true },
    ],
  },
}));

describe("SignupPage", () => {
  beforeEach(() => {
    mockSignupFn.mockClear();
  });

  it("renders the sign-up page correctly", () => {
    renderWithProviders(<SignupPage />);
    expect(screen.getByTestId("signup-page-container")).toMatchSnapshot();
  });

  it("submits the form with name, email, and password", async () => {
    renderWithProviders(<SignupPage />);

    fireEvent.change(screen.getByLabelText(/name/i), {
      target: { value: "Test User" },
    });
    fireEvent.change(screen.getByLabelText(/email/i), {
      target: { value: "test@example.com" },
    });
    fireEvent.change(screen.getByLabelText(/password/i), {
      target: { value: "password123!" },
    });

    fireEvent.click(screen.getByRole("button", { name: /sign up/i }));

    await waitFor(() => expect(mockSignupFn).toHaveBeenCalledTimes(1));

    expect(mockSignupFn).toHaveBeenCalledWith({
      name: "Test User",
      email: "test@example.com",
      password: "password123!",
    });
  });
});
