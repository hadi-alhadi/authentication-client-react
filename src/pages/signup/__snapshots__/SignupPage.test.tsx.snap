// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignupPage renders the sign-up page correctly 1`] = `
<div
  class="sc-beySPh cIGrgw"
  data-testid="signup-page-container"
>
  <div
    class="sc-guDLey fbkvNj card"
  >
    <div
      class="sc-dmyCSP kWnFwo card-title h5"
    >
      Sign Up Page
    </div>
    <form
      class="sc-hLQSwg cCqvVm"
      novalidate=""
    >
      <div
        class="col-md-12"
      >
        <label
          class="form-label"
          for="validationFormikName"
        >
          Name
        </label>
        <div
          class="input-group has-validation"
        >
          <input
            aria-describedby="inputGroupPrepend"
            class="form-control"
            id="validationFormikName"
            name="name"
            placeholder="Your Name"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="col-md-12"
      >
        <label
          class="form-label"
          for="validationFormikEmail"
        >
          Email
        </label>
        <div
          class="input-group has-validation"
        >
          <input
            aria-describedby="inputGroupPrepend"
            class="form-control"
            id="validationFormikEmail"
            name="email"
            placeholder="Your email address"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="col-md-12"
      >
        <label
          class="form-label"
          for="validationFormikPassword"
        >
          Password
        </label>
        <div
          class="input-group has-validation"
        >
          <input
            aria-describedby="inputGroupPrepend"
            class="form-control"
            id="validationFormikPassword"
            name="password"
            placeholder="Your Password"
            type="password"
            value=""
          />
        </div>
      </div>
      <div>
        <span>
          Already have an account?
        </span>
         
        <a
          href="/signin"
        >
          Sign In
        </a>
      </div>
      <button
        class="sc-eDLKkx jYVFEp"
        type="submit"
      >
        Sign Up
      </button>
    </form>
  </div>
</div>
`;
